components/asset-store-controller-manager/pkg/controller/clusterasset/clusterasset_controller.go

```diff
@@ -136,96 +149,105 @@ func (r *ReconcileClusterAsset) Reconcile(request reconcile.Request) (reconcile.
+func (r *ReconcileClusterAsset) removeFinalizer(ctx context.Context, namespacedName types.NamespacedName) error {
+	updateFnc := func(instance *assetstorev1alpha2.ClusterAsset) error {
+		if instance.DeletionTimestamp.IsZero() {
+			return nil
+		}
 
-	if err := r.updateStatus(ctx, instance, status); err != nil {
-		return reconcile.Result{}, err
+		copy := instance.DeepCopy()
+		r.finalizer.DeleteFrom(copy)
+
+		return r.Update(ctx, copy)
 	}
 
-	return reconcile.Result{RequeueAfter: r.relistInterval}, nil
+	return r.update(ctx, namespacedName, updateFnc)
 }
 
-func (r *ReconcileClusterAsset) onFailed(ctx context.Context, instance *assetstorev1alpha2.ClusterAsset) (reconcile.Result, error) {
-	status, err := r.handler.OnFailed(ctx, instance, instance.Spec.CommonAssetSpec, instance.Status.CommonAssetStatus)
-	if err != nil {
-		return reconcile.Result{}, err
-	}
+func (r *ReconcileClusterAsset) updateStatus(ctx context.Context, namespacedName types.NamespacedName, commonStatus *assetstorev1alpha2.CommonAssetStatus) error {
+	updateFnc := func(instance *assetstorev1alpha2.ClusterAsset) error {
+		if r.isStatusUnchanged(instance, commonStatus) {
+			return nil
+		}
 
-	if err := r.updateStatus(ctx, instance, *status); err != nil {
-		return reconcile.Result{}, err
+		copy := instance.DeepCopy()
+		copy.Status.CommonAssetStatus = *commonStatus
+
+		return r.Status().Update(ctx, copy)
 	}
 
-	return reconcile.Result{RequeueAfter: r.relistInterval}, nil
+	return r.update(ctx, namespacedName, updateFnc)
 }
 
-func (r *ReconcileClusterAsset) onAddOrUpdate(ctx context.Context, instance *assetstorev1alpha2.ClusterAsset) (reconcile.Result, error) {
-	if !r.finalizer.IsDefinedIn(instance) {
-		r.finalizer.AddTo(instance)
-		return reconcile.Result{Requeue: true}, r.Update(ctx, instance)
-	}
-	status := r.handler.OnAddOrUpdate(ctx, instance, instance.Spec.CommonAssetSpec, instance.Status.CommonAssetStatus)
-
-	if err := r.updateStatus(ctx, instance, status); err != nil {
-		return reconcile.Result{}, err
-	}
+func (r *ReconcileClusterAsset) isStatusUnchanged(instance *assetstorev1alpha2.ClusterAsset, newStatus *assetstorev1alpha2.CommonAssetStatus) bool {
+	currentStatus := instance.Status.CommonAssetStatus
 
-	return reconcile.Result{RequeueAfter: r.relistInterval}, nil
+	return newStatus == nil ||
+		currentStatus.ObservedGeneration == newStatus.ObservedGeneration &&
+			currentStatus.Phase == newStatus.Phase &&
+			currentStatus.Reason == newStatus.Reason
 }
 
-func (r *ReconcileClusterAsset) updateStatus(ctx context.Context, instance *assetstorev1alpha2.ClusterAsset, commonStatus assetstorev1alpha2.CommonAssetStatus) error {
-	toUpdate := instance.DeepCopy()
-	toUpdate.Status.CommonAssetStatus = commonStatus
+func (r *ReconcileClusterAsset) update(ctx context.Context, namespacedName types.NamespacedName, updateFnc func(instance *assetstorev1alpha2.ClusterAsset) error) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
I would rename this function to `retry.OnConflict()`

Reply:
Sure, I can contribute too `k8s.io/client-go` 😄

Reply:
😂 Okay, now I can see the import